internal/config/seccomp/notifier.go

```diff
@@ -0,0 +1,370 @@
+		return nil, nil
+	}
+	if _, ok := sandboxAnnotations[annotations.SeccompNotifierActionAnnotation]; !ok {
+		return nil, nil
+	}
+
+	log.Infof(ctx, "Injecting seccomp notifier into seccomp profile of container %s", containerID)
+
+	isActionToOverride := func(action specs.LinuxSeccompAction) bool {
+		if action == specs.ActErrno ||
+			action == specs.ActKill ||
+			action == specs.ActKillProcess ||
+			action == specs.ActKillThread {
+			return true
+		}
+
+		return false
+	}
+
+	if isActionToOverride(profile.DefaultAction) {
+		log.Infof(
+			ctx,
+			"The seccomp profile default action %s cannot be overridden to %s, "+
+				"which means that syscalls using that default action can't be "+
+				"traced by the notifier",
+			profile.DefaultAction, seccomp.ActNotify,
+		)
+	}
+
+	for i, syscall := range profile.Syscalls {
+		if isActionToOverride(syscall.Action) {
+			profile.Syscalls[i].Action = specs.ActNotify
+		}
+	}
+
+	profile.ListenerPath = filepath.Join(c.NotifierPath(), containerID)
+
+	notifier, err := NewNotifier(ctx, msgChan, containerID, profile.ListenerPath, sandboxAnnotations)
+	if err != nil {
+		return nil, fmt.Errorf("unable to run notifier: %w", err)
+	}
+
+	return notifier, nil
+}
+
+// NewNotifier starts the notifier for the provided arguments.
+func NewNotifier(
+	ctx context.Context,
+	msgChan chan Notification,
+	containerID, listenerPath string,
+	annotationMap map[string]string,
+) (*Notifier, error) {
+	log.Infof(ctx, "Waiting for seccomp file descriptor on container %s", containerID)
+	listener, err := net.Listen("unix", listenerPath)
+	if err != nil {
+		return nil, fmt.Errorf("listen for seccomp socket: %w", err)
+	}
+
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if errors.Is(err, net.ErrClosed) {
+				log.Infof(ctx, "Stopping notifier for container %s", containerID)
+				return
+			}
+			if err != nil {
+				log.Errorf(ctx, "Unable to accept connection: %v", err)
+				continue
+			}
+			log.Debugf(ctx, "Got new seccomp notifier connection")
+
+			socket, err := conn.(*net.UnixConn).File()
+			if err := conn.Close(); err != nil {
+				log.Errorf(ctx, "Unable to close connection: %v", err)
+				continue
+			}
+			if err != nil {
+				log.Errorf(ctx, "Unable to get socket: %v", err)
+				continue
+			}
+
+			newFd, err := handleNewMessage(int(socket.Fd()))
+			if err := socket.Close(); err != nil {
+				log.Errorf(ctx, "Unable to close socket: %v", err)
+				continue
+			}
+			if err != nil {
+				log.Errorf(ctx, "Unable to receive seccomp file descriptor: %v", err)
+				continue
+			}
+
+			log.Infof(ctx, "Received new seccomp fd: %v", newFd)
+			go handler(ctx, containerID, msgChan, libseccomp.ScmpFd(newFd))
+		}
+	}()
+
+	action, ok := annotationMap[annotations.SeccompNotifierActionAnnotation]
+	if !ok {
+		return nil, fmt.Errorf("%s annotation not set on container", annotations.SeccompNotifierActionAnnotation)
+	}
+
+	const actionStop = "stop"
```

Comment:
WDYT about putting this in `pkg/annotations` or some similar public place? I know we document the value but it also seems worth allowing folks to vendor it. I'm 50/50 so I am fine as a followup or fine to not take that nit

Reply:
Good thought, changed it as suggested :+1: